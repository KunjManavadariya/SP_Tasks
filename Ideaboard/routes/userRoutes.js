import koaRouter from "koa-router";
import {
  verifyUser,
  verifyLogin,
  verifyUserDuplicates,
  verifyUserRegex,
} from "../middlewares";
const router = new koaRouter();
import {
  allUsers,
  deleteUser,
  loginUser,
  newUser,
  updateUser,
} from "../controllers/userControllers";

router.get("/users", verifyUser, allUsers);

router.post("/register", verifyUserDuplicates, verifyUserRegex, newUser);

router.post("/login", verifyLogin, loginUser);

router.delete("/user/:id", verifyUser, deleteUser);

router.patch("/:id", verifyUser, verifyUserRegex, updateUser);

export { router };
