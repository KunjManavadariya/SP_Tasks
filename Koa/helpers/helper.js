import axios from 'axios';
import env from 'dotenv';
env.config();
export const config = {
  headers: {
    Authorization: `Bearer ${process.env.bearer}`,
  },
};

export const helper = async function (hashtag) {
  try {
    const link = `https://api.twitter.com/2/tweets/search/recent?query=${encodeURIComponent(
      '#' + hashtag
    )}&tweet.fields=created_at,possibly_sensitive&expansions=author_id,attachments.media_keys&media.fields=media_key,preview_image_url,public_metrics,type,url`;
    const { data } = await axios.get(link, config);
    return data;
  } catch (error) {
    return { error };
  }
};
