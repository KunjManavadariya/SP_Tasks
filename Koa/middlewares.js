import { handleError } from './errors/errorHandler';
export const validateHashtag = async function (ctx, next) {
  const hashtag = ctx.request.params.hashtag;
  const hashtagValidator = /^[a-zA-Z0-9_]{1,280}$/;
  if (!hashtagValidator.test(hashtag)) {
    handleError(
      ctx,
      ctx.status,
      'Error in Hashtag. Enter a valid hashtag without spaces, hashtag or punctuation'
    );
    return;
  }
  await next();
};
