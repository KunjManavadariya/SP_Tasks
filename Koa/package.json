{
  "name": "koa",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "nodemon --exec babel-node ./index.js",
    "test": "jest --watch"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@babel/cli": "^7.20.7",
    "@babel/core": "^7.20.12",
    "@babel/node": "^7.20.7",
    "@babel/preset-env": "^7.20.2",
    "axios": "^1.2.3",
    "dotenv": "^16.0.3",
    "koa-bodyparser": "^4.3.0",
    "koa-router": "^12.0.0",
    "nodemon": "^2.0.20"
  },
  "devDependencies": {
    "jest": "^29.4.0"
  }
}
