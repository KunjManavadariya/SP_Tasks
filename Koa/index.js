import koaBodyParser from 'koa-bodyparser';
import Koa from 'koa';
import { router } from './routes/routes.js';

const app = new Koa();
const port = 3000;

//Routers Middlewares
app.use(koaBodyParser()).use(router.routes()).use(router.allowedMethods());
app.listen(port, () => {
  console.log(`Listening on ${port}`);
});
