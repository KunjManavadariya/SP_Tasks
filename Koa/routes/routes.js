import koaRouter from 'koa-router';
import { validateHashtag } from '../middlewares';
const router = new koaRouter();
import { home, searchHashtag, defaultRoute } from '../controllers/controller';

router.get('/', home);

router.get('/hashtag/:hashtag', validateHashtag, searchHashtag);

router.all(/.*/, defaultRoute);

export { router };
