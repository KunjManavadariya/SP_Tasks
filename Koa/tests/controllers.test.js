import request from 'supertest';
import Koa from 'koa';
import { router } from '../routes/routes';
import { helper } from '../helpers/helper';

describe('Testing routes', () => {
  let server;
  beforeAll(async () => {
    const app = new Koa();
    app.use(router.routes());
    app.use(router.allowedMethods());
    server = app.listen();
  });

  afterAll(() => {
    server.close();
  });

  test('GET /, returns an object', async () => {
    const response = await request(server).get('/').set('Accept', 'application/json');
    expect(response.status).toBe(200);
    expect(response.body).toEqual({ msg: 'Hello World' });
  });

  test('GET /hashtag:tag', async () => {
    const response = await request(server)
      .get('/hashtag/google')
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200);
    expect(response.body.latest.forEach(ele => ele.url));
  });

  test('Helper Axios testing', async () => {
    const response = [
      await helper('google'),
      await helper('#google'),
      await helper('ðŸ”¥'),
      await helper('goog  le'),
      await helper('goog@123'),
      await helper('ssjsdfhbakjhsdbakcvsjdhavjhvbdkjhv'),
    ];
    response.slice(1).forEach(ele => {
      expect(ele).toMatchSnapshot();
    });
    expect(response[0].error).not.toBeDefined();
    // expect(response[1].error).toBeDefined();
    // expect(response[2].error).toBeDefined();
    // expect(response[3].meta && response[3].meta.result_count).toBe(0);
    // expect(response[4].meta && response[4].meta.result_count).toBe(0);
    // expect(response[5].meta && response[5].meta.result_count).toBe(0);
  });

  test('Default route', async () => {
    const response = await request(server).get('/dckjsdhvb').set('Accept', 'application/json');
    expect(response.body).toMatchSnapshot();
    // expect(response.body.success).toEqual(false);
    // expect(response.status).toEqual(404);
  });

  test('Error in hashtags', async () => {
    const response = await request(server)
      .get('/hashtag/googl@123')
      .set('Accept', 'application/json');
    expect(response.body).toMatchSnapshot();
    // expect(response.body.success).toEqual(false);
    // expect(response.status).toEqual(404);
  });

  test('Gibrish hashtags', async () => {
    const response = await request(server)
      .get('/hashtag/googlealjhfHSDBCKSDH')
      .set('Accept', 'application/json');
    expect(response.body).toMatchSnapshot();
    // expect(response.body.success).toEqual(false);
    // expect(response.status).toEqual(404);
  });
});
