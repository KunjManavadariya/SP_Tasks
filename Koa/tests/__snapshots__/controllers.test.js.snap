// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing routes Default route 1`] = `
{
  "message": "Error in URL, no page found at the given URL. Check hashtag and enter proper URL again!",
  "status": 404,
  "success": false,
}
`;

exports[`Testing routes Error in hashtags 1`] = `
{
  "message": "Error in Hashtag. Enter a valid hashtag without spaces, hashtag or punctuation",
  "status": 404,
  "success": false,
}
`;

exports[`Testing routes Gibrish hashtags 1`] = `
{
  "message": "No tweet found with the given hashtag. Try giving a more commonly used hashtag.",
  "status": 404,
  "success": false,
}
`;

exports[`Testing routes Helper Axios testing 1`] = `
{
  "error": [AxiosError: Request failed with status code 400],
}
`;

exports[`Testing routes Helper Axios testing 2`] = `
{
  "error": [AxiosError: Request failed with status code 503],
}
`;

exports[`Testing routes Helper Axios testing 3`] = `
{
  "meta": {
    "result_count": 0,
  },
}
`;

exports[`Testing routes Helper Axios testing 4`] = `
{
  "meta": {
    "result_count": 0,
  },
}
`;

exports[`Testing routes Helper Axios testing 5`] = `
{
  "meta": {
    "result_count": 0,
  },
}
`;
